### PR TITLE
Allow "any" type for Express.Request.user in Passport

### PR DESCRIPTION
**Pull request by joshcanhelp**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code.
- [x] Add or edit tests to reflect the change (no tests were added or edited in the change that this reverts. Tests _may_ have prevented this issue in the first place so I'm open to suggestions here ... `npm test` ran and passed.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint passport`

**Changing an existing definition:**

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: auth0/node-jwks-rsa#47 and DefinitelyTyped/DefinitelyTyped#23976 report the issue, was created in [this commit](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/c75ee379eb78e44d3c7a04d0ab4f10ed93625c1d)
- [ ] Increase the version number in the header if appropriate (not sure if this is necessary)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by tlaziuk — APPROVED** [review]

**Comment by hscasn** [comment]
@tlaziuk Why is this an improvement? Before we could just extend the interface User in another .d.ts and keep type safety like this:

```
declare global {
  export namespace Express {
    export interface User extends UserModel {}
    export interface Request {
      user?: User;
    }
  }
}
```

But now there are two problems:
1- This is a breaking change. After updating to the latest version my code does not compile anymore, since I can not reassign "any" as "User".
2- Types for `req.user` will always be "any" now and we lose type safety

**Comment by joshcanhelp** [comment]
@hscasn - I can comment on the reasoning here. 

I added 2 issues in the PR description that were caused by the change that this reverts. While I understand that this defeats the purpose of type safety, the original change caused breaking changes with existing definitions. I don't know that there is a great way to define types for global namespaces without causing collisions. 

I'm curious though ... why would this cause a compilation error on your end? The `any` type should allow, well, anything.

**Comment by hscasn** [comment]
@joshcanhelp Instead of "any", maybe we could define the type as "user?: User | object | string..."  or something like this.

The compilation error happens because in my case I am merging the declaration of the "User" interface: the interface built into Passport is a generic object, and I am specifying exactly what properties that object has:

```
interface A {
    [idx: string]: any;
}

// Legal: propA keeps its type, and now it knows that it has one property called "propA" which is a number
interface A {
    propA: number;
}
```

The new version, however, says that "User" is of any type, which I can not merge.
```
interface B {
    propA: any;
}

// Illegal: propA must have the same type, and "{ propA: number }" is not assignable to "any"
interface B {
    propA: number;
}
```

**Comment by Bene-Graham** [comment]
Just want to also comment that this has also broken my builds. Since I don't allow any to be used.

In the previous version this is what I did to extend the User interface.

```
declare global {
    namespace Express {
        // tslint:disable-next-line interface-name no-empty-interface
        interface User extends IUser {
        }
    }
}
```

Where IUser is a internal interface with my applications specific properties.

**Comment by wcauchois** [comment]
I was wondering as well if there was any way to provide a strong, custom, type for `req.user` for my application.

I added

```typescript
declare namespace Express {
  interface Request {
    user?: { /* custom properties etc */ };
  }
}
```

to an express.d.ts but then I get this error:

```
node_modules/@types/passport/index.d.ts:15:13 - error TS2717: Subsequent property declarations must have the same type.  Property 'user' must be of type '<my type, redacted>', but here has type 'any'.

15             user?: any;
```

Is there any way to provide a strong type for `user`? This seems like a pretty inconvenient breach of type safety.

Thanks for any help or thoughts anyone can provide.

**Comment by Bene-Graham** [comment]
As far as I am aware the only way to get type safety back is to go back to what it was doing earlier
```TS
 namespace Express {
        interface Request {
            user?: User;
        }
        interface User {
            [_: string]: any;
        }
    }
}
```

**Comment by danielpa9708** [comment]
So, is there a way to make it work like before and not break the current code, maybe increment the major version ?

File: types/passport/index.d.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ declare global {
     namespace Express {
         interface Request {
             authInfo?: any;
-            user?: User;
+            user?: any;
```
Comment (joeandaverde) [inline types/passport/index.d.ts]: What's the point of types if it's just an any? This should be generic if at all possible.

**Comment by joeandaverde** [comment]
I'm broken on this too.

**Pull request merged**